compiler/rustc_mir_transform/src/simplify_static_switch.rs

```diff
@@ -0,0 +1,165 @@
+use rustc_mir_dataflow::impls::borrowed_locals;
+use rustc_session::Session;
+
+use super::simplify::remove_dead_blocks;
+
+/// This pass looks to optimize a pattern in MIR where variants of an aggregate
+/// are constructed in one or more blocks with the same successor and then that
+/// aggregate/discriminant is switched on in that successor block, in which case
+/// we can remove the switch on the discriminant because we statically know
+/// what target block will be taken for each variant.
+///
+/// Note that an aggregate which is returned from a function call or passed as
+/// an argument is not viable for this optimization because we do not statically
+/// know the discriminant/variant of the aggregate.
+///
+/// For example, the following CFG:
+/// ```text
+///     x = Foo::A(y); ---       Foo::A ---> ...
+///    /                  \            /
+/// ...                    --> switch x
+///    \                  /            \
+///     x = Foo::B(y); ---       Foo::B ---> ...
+/// ```
+/// would become:
+/// ```text
+///     x = Foo::A(y); --------- Foo::A ---> ...
+///    /
+/// ...
+///    \
+///     x = Foo::B(y); --------- Foo::B ---> ...
+/// ```
+pub struct SimplifyStaticSwitch;
+
+impl<'tcx> MirPass<'tcx> for SimplifyStaticSwitch {
+    fn name(&self) -> &'static str {
+        "SimplifyStaticSwitch"
+    }
+
+    fn is_enabled(&self, sess: &Session) -> bool {
+        sess.mir_opt_level() >= 2
+    }
+
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        debug!(body = ?body.source.def_id());
+        if simplify_static_switch(tcx, body) {
+            remove_dead_blocks(tcx, body);
+        }
+    }
+}
+
+#[instrument(level = "debug", skip(tcx, body), ret)]
+fn simplify_static_switch<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) -> bool {
+    let predecessors = body.basic_blocks.predecessors();
+    let borrowed_locals = borrowed_locals(body);
+    let mut new_targets = FxHashMap::default();
+    for (block, data) in body.basic_blocks.iter_enumerated() {
+        if data.is_cleanup {
+            continue;
+        }
+
+        let predecessors = &predecessors[block];
+        if predecessors.is_empty() {
+            continue;
+        }
+
+        let TerminatorKind::SwitchInt {
+            discr: Operand::Move(switched) | Operand::Copy(switched),
+            ref targets
+        } = data.terminator().kind else {
+            continue
+        };
+
+        let Some(switched) = switched.as_local() else {
+            continue
+        };
+
+        if borrowed_locals.contains(switched) {
+            continue;
+        }
+
+        let Some(discr) = data.statements.iter().rev().find_map(|statement| {
+            match statement.kind {
+                StatementKind::Assign(box (place, Rvalue::Discriminant(discr)))
+                    if place.as_local() == Some(switched)
+                        && let Some(local) = discr.as_local()
+                        && !borrowed_locals.contains(local) =>
```

Comment:
Those checks should happen outside of the `find_map`, in cases there are several assignments to `switched`.

Reply:
i also moved these checks to outside of the `find_map`.